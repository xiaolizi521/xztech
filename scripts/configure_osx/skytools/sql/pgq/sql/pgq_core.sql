
select * from pgq.maint_tables_to_vacuum();
select * from pgq.maint_retry_events();

select pgq.create_queue('tmpqueue');
select pgq.register_consumer('tmpqueue', 'consumer');
select pgq.unregister_consumer('tmpqueue', 'consumer');
select pgq.drop_queue('tmpqueue');

select pgq.create_queue('myqueue');
select pgq.register_consumer('myqueue', 'consumer');
select pgq.next_batch('myqueue', 'consumer');
select pgq.next_batch('myqueue', 'consumer');
select pgq.ticker();
select pgq.next_batch('myqueue', 'consumer');
select pgq.next_batch('myqueue', 'consumer');

select queue_name, consumer_name, prev_tick_id, tick_id, lag from pgq.get_batch_info(1);

select queue_name, queue_ntables, queue_cur_table, queue_rotation_period,
       queue_switch_time <= now() as switch_time_exists,
       queue_external_ticker, queue_ticker_max_count, queue_ticker_max_lag,
       queue_ticker_idle_period, ticker_lag < '2 hours' as ticker_lag_exists
  from pgq.get_queue_info() order by 1;
select queue_name, consumer_name, lag < '30 seconds' as lag_exists,
       last_seen < '30 seconds' as last_seen_exists,
       last_tick, current_batch, next_tick
  from pgq.get_consumer_info() order by 1, 2;

select pgq.finish_batch(1);
select pgq.finish_batch(1);

select pgq.ticker();
select pgq.next_batch('myqueue', 'consumer');
select * from pgq.batch_event_tables(2);
select * from pgq.get_batch_events(2);
select pgq.finish_batch(2);

select pgq.insert_event('myqueue', 'r1', 'data');
select pgq.insert_event('myqueue', 'r2', 'data', 'extra1', 'extra2', 'extra3', 'extra4');
select pgq.insert_event('myqueue', 'r3', 'data');
select pgq.current_event_table('myqueue');
select pgq.ticker();

select pgq.next_batch('myqueue', 'consumer');
select ev_id,ev_retry,ev_type,ev_data,ev_extra1,ev_extra2,ev_extra3,ev_extra4 from pgq.get_batch_events(3);

select * from pgq.failed_event_list('myqueue', 'consumer');

select pgq.event_failed(3, 1, 'failure test');
select pgq.event_failed(3, 1, 'failure test');
select pgq.event_retry(3, 2, 0);
select pgq.event_retry(3, 2, 0);
select pgq.finish_batch(3);

select ev_failed_reason, ev_id, ev_txid, ev_retry, ev_type, ev_data
  from pgq.failed_event_list('myqueue', 'consumer');

select ev_failed_reason, ev_id, ev_txid, ev_retry, ev_type, ev_data
  from pgq.failed_event_list('myqueue', 'consumer', 0, 1);

select * from pgq.failed_event_count('myqueue', 'consumer');
select * from pgq.failed_event_delete('myqueue', 'consumer', 0);

select pgq.event_retry_raw('myqueue', 'consumer', now(), 666, now(), 0,
        'rawtest', 'data', null, null, null, null);

select pgq.ticker();

-- test maint
update pgq.queue set queue_rotation_period = '0 seconds';
select queue_name, pgq.maint_rotate_tables_step1(queue_name) from pgq.queue;
select pgq.maint_rotate_tables_step2();

-- test extra
select nextval(queue_event_seq) from pgq.queue where queue_name = 'myqueue';
select pgq.force_tick('myqueue');
select nextval(queue_event_seq) from pgq.queue where queue_name = 'myqueue';

