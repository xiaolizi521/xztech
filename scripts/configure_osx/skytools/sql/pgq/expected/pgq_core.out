select * from pgq.maint_tables_to_vacuum();
 maint_tables_to_vacuum 
------------------------
 pgq.subscription
 pgq.consumer
 pgq.queue
 pgq.tick
 pgq.retry_queue
(5 rows)

select * from pgq.maint_retry_events();
 maint_retry_events 
--------------------
                  0
(1 row)

select pgq.create_queue('tmpqueue');
 create_queue 
--------------
            1
(1 row)

select pgq.register_consumer('tmpqueue', 'consumer');
 register_consumer 
-------------------
                 1
(1 row)

select pgq.unregister_consumer('tmpqueue', 'consumer');
 unregister_consumer 
---------------------
                   1
(1 row)

select pgq.drop_queue('tmpqueue');
 drop_queue 
------------
          1
(1 row)

select pgq.create_queue('myqueue');
 create_queue 
--------------
            1
(1 row)

select pgq.register_consumer('myqueue', 'consumer');
 register_consumer 
-------------------
                 1
(1 row)

select pgq.next_batch('myqueue', 'consumer');
 next_batch 
------------
           
(1 row)

select pgq.next_batch('myqueue', 'consumer');
 next_batch 
------------
           
(1 row)

select pgq.ticker();
 ticker 
--------
      1
(1 row)

select pgq.next_batch('myqueue', 'consumer');
 next_batch 
------------
          1
(1 row)

select pgq.next_batch('myqueue', 'consumer');
 next_batch 
------------
          1
(1 row)

select queue_name, consumer_name, prev_tick_id, tick_id, lag from pgq.get_batch_info(1);
 queue_name | consumer_name | prev_tick_id | tick_id |     lag     
------------+---------------+--------------+---------+-------------
 myqueue    | consumer      |            1 |       2 | @ 0.00 secs
(1 row)

select queue_name, queue_ntables, queue_cur_table, queue_rotation_period,
       queue_switch_time <= now() as switch_time_exists,
       queue_external_ticker, queue_ticker_max_count, queue_ticker_max_lag,
       queue_ticker_idle_period, ticker_lag < '2 hours' as ticker_lag_exists
  from pgq.get_queue_info() order by 1;
 queue_name | queue_ntables | queue_cur_table | queue_rotation_period | switch_time_exists | queue_external_ticker | queue_ticker_max_count | queue_ticker_max_lag | queue_ticker_idle_period | ticker_lag_exists 
------------+---------------+-----------------+-----------------------+--------------------+-----------------------+------------------------+----------------------+--------------------------+-------------------
 myqueue    |             3 |               0 | @ 2 hours             | t                  | f                     |                    500 | @ 3 secs             | @ 1 min                  | t
(1 row)

select queue_name, consumer_name, lag < '30 seconds' as lag_exists,
       last_seen < '30 seconds' as last_seen_exists,
       last_tick, current_batch, next_tick
  from pgq.get_consumer_info() order by 1, 2;
 queue_name | consumer_name | lag_exists | last_seen_exists | last_tick | current_batch | next_tick 
------------+---------------+------------+------------------+-----------+---------------+-----------
 myqueue    | consumer      | t          | t                |         1 |             1 |         2
(1 row)

select pgq.finish_batch(1);
 finish_batch 
--------------
            1
(1 row)

select pgq.finish_batch(1);
WARNING:  finish_batch: batch 1 not found
 finish_batch 
--------------
            0
(1 row)

select pgq.ticker();
 ticker 
--------
      1
(1 row)

select pgq.next_batch('myqueue', 'consumer');
 next_batch 
------------
          2
(1 row)

select * from pgq.batch_event_tables(2);
 batch_event_tables 
--------------------
 pgq.event_2_0
(1 row)

select * from pgq.get_batch_events(2);
 ev_id | ev_time | ev_txid | ev_retry | ev_type | ev_data | ev_extra1 | ev_extra2 | ev_extra3 | ev_extra4 
-------+---------+---------+----------+---------+---------+-----------+-----------+-----------+-----------
(0 rows)

select pgq.finish_batch(2);
 finish_batch 
--------------
            1
(1 row)

select pgq.insert_event('myqueue', 'r1', 'data');
 insert_event 
--------------
            1
(1 row)

select pgq.insert_event('myqueue', 'r2', 'data', 'extra1', 'extra2', 'extra3', 'extra4');
 insert_event 
--------------
            2
(1 row)

select pgq.insert_event('myqueue', 'r3', 'data');
 insert_event 
--------------
            3
(1 row)

select pgq.current_event_table('myqueue');
 current_event_table 
---------------------
 pgq.event_2_0
(1 row)

select pgq.ticker();
 ticker 
--------
      1
(1 row)

select pgq.next_batch('myqueue', 'consumer');
 next_batch 
------------
          3
(1 row)

select ev_id,ev_retry,ev_type,ev_data,ev_extra1,ev_extra2,ev_extra3,ev_extra4 from pgq.get_batch_events(3);
 ev_id | ev_retry | ev_type | ev_data | ev_extra1 | ev_extra2 | ev_extra3 | ev_extra4 
-------+----------+---------+---------+-----------+-----------+-----------+-----------
     1 |          | r1      | data    |           |           |           | 
     2 |          | r2      | data    | extra1    | extra2    | extra3    | extra4
     3 |          | r3      | data    |           |           |           | 
(3 rows)

select * from pgq.failed_event_list('myqueue', 'consumer');
 ev_failed_reason | ev_failed_time | ev_id | ev_time | ev_txid | ev_owner | ev_retry | ev_type | ev_data | ev_extra1 | ev_extra2 | ev_extra3 | ev_extra4 
------------------+----------------+-------+---------+---------+----------+----------+---------+---------+-----------+-----------+-----------+-----------
(0 rows)

select pgq.event_failed(3, 1, 'failure test');
 event_failed 
--------------
            1
(1 row)

select pgq.event_failed(3, 1, 'failure test');
 event_failed 
--------------
            0
(1 row)

select pgq.event_retry(3, 2, 0);
 event_retry 
-------------
           1
(1 row)

select pgq.event_retry(3, 2, 0);
 event_retry 
-------------
           0
(1 row)

select pgq.finish_batch(3);
 finish_batch 
--------------
            1
(1 row)

select ev_failed_reason, ev_id, ev_txid, ev_retry, ev_type, ev_data
  from pgq.failed_event_list('myqueue', 'consumer');
 ev_failed_reason | ev_id | ev_txid | ev_retry | ev_type | ev_data 
------------------+-------+---------+----------+---------+---------
 failure test     |     1 |         |        0 | r1      | data
(1 row)

select ev_failed_reason, ev_id, ev_txid, ev_retry, ev_type, ev_data
  from pgq.failed_event_list('myqueue', 'consumer', 0, 1);
 ev_failed_reason | ev_id | ev_txid | ev_retry | ev_type | ev_data 
------------------+-------+---------+----------+---------+---------
(0 rows)

select * from pgq.failed_event_count('myqueue', 'consumer');
 failed_event_count 
--------------------
                  1
(1 row)

select * from pgq.failed_event_delete('myqueue', 'consumer', 0);
ERROR:  event not found
select pgq.event_retry_raw('myqueue', 'consumer', now(), 666, now(), 0,
        'rawtest', 'data', null, null, null, null);
 event_retry_raw 
-----------------
             666
(1 row)

select pgq.ticker();
 ticker 
--------
      1
(1 row)

-- test maint
update pgq.queue set queue_rotation_period = '0 seconds';
select queue_name, pgq.maint_rotate_tables_step1(queue_name) from pgq.queue;
 queue_name | maint_rotate_tables_step1 
------------+---------------------------
 myqueue    |                         1
(1 row)

select pgq.maint_rotate_tables_step2();
 maint_rotate_tables_step2 
---------------------------
                         1
(1 row)

-- test extra
select nextval(queue_event_seq) from pgq.queue where queue_name = 'myqueue';
 nextval 
---------
       4
(1 row)

select pgq.force_tick('myqueue');
 force_tick 
------------
          5
(1 row)

select nextval(queue_event_seq) from pgq.queue where queue_name = 'myqueue';
 nextval 
---------
    1006
(1 row)

